### ISSUE TITLE
Using `get` with a cluster

### ISSUE THREAD

**Issue by jakirkham** [ContinuumIO/dask#438]
So, I have just looked over the cluster documentation. I am trying to implement something simple. For instance, say I have a `dask.array` (`a`) that I would like to take the max over one dimension (e.g. the `0`-th), which can be done simply to give me a new `dask.array` (`am`). Normally, if I want to get this value, I could allocate an array and store it. In the case of a dask client, I am not sure how to proceed. 

    >>> am = a.max(axis=0)

**Comment by mrocklin** [comment]
If you call `.compute(get=myclient.get)` on an array then the result will be sent back from the workers to the client.

```python
myclient = Client('tcp://scheduler-hostname:1234')
result = am.compute(get=myclient.get)
```

If you want to store the result to something like an HDF5 file then all of the workers must have access to that thing, presumably through some network file system.  This is a bit tricky because each worker will have to open the HDF5 file on their own (h5py.datasets aren't serializable).  At the moment there isn't a nice convenience function accomplish this.  There should be though.

**Comment by jakirkham** [comment]
Ah, ok. Thanks for clarifying.

> If you want to store the result to something like an HDF5 file then all of the workers must have access to that thing, presumably through some network file system.

Is this assuming that one has Parallel HDF5 then?

> This is a bit tricky because each worker will have to open the HDF5 file on their own (h5py.datasets aren't serializable)

Indeed. So, if the `dask.array` was using `h5py.Dataset` to read from, would this also cause issues?

**Comment by jakirkham** [comment]
> This is a bit tricky because each worker will have to open the HDF5 file on their own (h5py.datasets aren't serializable)

Another thought on this. Maybe you could have something like, but admittedly not the same as, iPython's `%px`/`%%px` where it runs the same command on all engines. As you can't really use cell magics, maybe a `with` block would be an option.

**Comment by mrocklin** [comment]
I would expect `from_array(h5py.File(...)['/data'])` to fail in a parallel setting where the dataset needed to be communicated across a wire.  I'm surprised that this doesn't fail for you.

I'm not sure exactly how safe HDF5 is in parallel.  My guess is that it's decently safe for reads even if not compiled specially.

**Comment by jakirkham** [comment]
> I would expect from_array(h5py.File(...)['/data']) to fail in a parallel setting where the dataset needed to be communicated across a wire. I'm surprised that this doesn't fail for you.

Never mind, I was doing something wrong. It doesn't seem to work. However, it doesn't explicitly fail, but just hangs.

> I'm not sure exactly how safe HDF5 is in parallel. My guess is that it's decently safe for reads even if not compiled specially.

Yes, reads are perfectly fine in parallel, but writes are not. I was musing more over the problem of how to store to HDF5 when running in the distributed mode.

**Comment by mrocklin** [comment]
Hrm, so how would you normally solve this problem in practice without `dask`?  Parallel HDF5?  Some other solution?

**Comment by jakirkham** [comment]
Which part?

If you are referring to reading, I bet serializing the filename and dataset name would be sufficient and then a file can be opened in read mode on the engines. However, I am sure you already thought about this. :)


Regarding the second, there are a few ways.

One is Parallel HDF5. However, this requires MPI and a special build of HDF5 and h5py. This could be a way to go. So, it could be reasonable, but these dependency constraints might not be nice.

If you skip this route there are two ways to go, one would be to have some main process control writing to the HDF5 file. Then the engines serializes their data over the wire and the main process handles putting this to disk. The main process in this case would likely be the client or some other entity not yet included in your model.

Another solution, is have each engine write an HDF5 file of their own. Then the client can stitch them together. This is a pretty simple solution to implement and avoids heavy traffic.

**Comment by jakirkham** [comment]
IMHO, the last case (having each process write to a file is easy, safe, reasonably scalable, and will work without adding more dependencies) is probably best.

**Comment by jakirkham** [comment]
How does serialization of arrays to the engines currently work? Does the scheduler serialize the whole array or does it simply hook up some sort of connection between the client and relevant engines?

**Comment by mrocklin** [comment]
The graph moves from the client to the scheduler, the scheduler moves tasks of the graph to the workers as necessary.  When workers require intermediate results from each other then they communicate peer-to-peer.

Ideally the graph itself is relatively small and does not hold large data like arrays, but instead holds functions that say how to get an array (e.g. by loading from a network file system).

**Comment by mrocklin** [comment]
@jakirkham do you have any further questions on this topic or can this issue be closed?

**Comment by jakirkham** [comment]
Sorry to drop off the face of the map. I needed a solution quickly and had to cobble one together. This is what I have learned.

Doing HDF5 I/O becomes a major bottleneck if it is happening on the client side. That being said, it is much more tolerable if **reading** is happening on **engines**. This way, no matter how bad your I/O situation is whether you are using SSD, HD, or NFS, the engines will remain active and won't be idling until data comes through the pipe. This is very obvious on simple operations like element-wise multiplication. However, it remained quite noticeable for fairly complex operations.

As we already know, HDF5 only allows one write handle. So, it is natural to place this on the client-side. Also, as reading and writing is not [*yet*]( http://docs.h5py.org/en/latest/swmr.html ) available it will have to be a different file from the input data. This actually works ok as now there is 1 HDF5 file per array. If they really need to be on the same file at the end, a temporary file can be used and a copy step to the original file can be added.

As mentioned, I/O still can be an issue when there is one process doing it. However, if jobs have already been queued up in advance, this becomes less of an issue. Given a comparable amount of work is allocated to each engine and a significantly large amount of work, engines will over time have some small lag between them. This results in parallelism that is somewhat less than the number of engines, but is not 1.

In my experiences, I have found this to be acceptable.
